Add section on Fortran rewrite (#280)
scipy/scipy-articles: scipy-2.0/src/paper.tex

\section{Introduction}

\section{A New Look at Language Choices}

As discussed in~\cite{Virtanen2020}, much of SciPy's original success can be attributed
to the convenient Python interfaces it provided to popular compiled numerical libraries
that were otherwise distributed primarily as standalone Fortran source
packages, such as those available through the Netlib repository~\cite{netlib}.

> This included:

Reviewer: ```suggestion
These included
```

ODEPACK~\cite{Hindmarsh1983ODEPACK},
VODE/ZVODE~\cite{Brown1989}, and DOPRI5~\cite{Hairer1987} for solving initial value
problems; QUADPACK~\cite{Piessens1983} for quadrature; FITPACK~\cite{dierckx1995curve}
for spline fitting; ID~\cite{martinsson2008id,Halko2011} for interpolative
matrix decomposition; ODRPACK~\cite{Boggs1989}
for orthogonal distance regression; COBYLA~\cite{Powell1994}, L-BFGS-B~\cite{Morales2011},
SLSQP~\cite{Kraft1994,Kraft1988SQP}, MINPACK~\cite{More1980}, and
NNLS~\cite{Lawson1995} for optimization;
ARPACK-NG~\cite{Lehoucq2023ARPACKNG} and PROPACK~\cite{Larsen1998} for sparse eigenvalue problems

> and singular value decompositions respectively; AMOS~\cite{Amos1986} and

Reviewer: ```suggestion
and singular value decompositions, respectively; AMOS~\cite{Amos1986} and
```

mature enough to be adopted by SciPy at the time and waiting for them to mature was
deemed not viable.

In addition, there were a number of further issues pertaining to the Fortran 77
implementations of these codes. A number of them were not thread-safe,
meaning that they could not be safely called from multiple threads simultaneously.
Previously, SciPy used locking to prevent this. However, the introduction of
free-threading in CPython 3.13, and its subsequent maturation in CPython 3.14, meant
that, for the first time, multiple threads could execute Python code concurrently
without the Global Interpreter Lock (GIL)~\cite{gross2023pep703}. Unfortunately,
users of these non-reentrant codes would not be able to take advantage of this
new feature.
Other challenges presented by these codes included a lack of support for
64-bit integer types (ILP64) and lack of control over the random number generator used.
Finally, many of these codes are no longer actively maintained, meaning that the
SciPy developers were effectively responsible for fixing any bugs that arose.
However, being a Python library, few members of the SciPy developer community had expertise

> to tackle these issues. This often led to bugs remaining unresolved indefinitely.

Reviewer: following some of the comments, what about?

```suggestion
This increased the maintenance burden and made it more difficult to resolve bugs and improve these routines.
```

Author: I wasn't sure if it was important to acknowledge that a lack of Fortran experience was also a factor. WDYT?

Reviewer: yes, we could mention it as suggested by @ev-br https://github.com/scipy/scipy-articles/pull/280#issuecomment-5233334599. The "maintenance burden" is sort of a broader version of that.

After careful evaluation of the above issues, it was decided that for SciPy 2.0, these
codes needed either to be replaced with
more recent, actively maintained alternatives, or to be translated from their
original Fortran 77 source code to either pure Python,

> Cython~\cite{Behnel2011}, or C/C++ in order to secure the long term future of the

Reviewer: ```suggestion
Cython~\cite{Behnel2011}, or C/C++ in order to secure the long-term future of the
```

\section{Introduction}

> \section{Modernisation of legacy compiled codes}

Author: maybe "legacy" isn't the right word

Reviewer 2: The original article had a section, `Language Choices`

<img width="850" height="265" alt="image" src="https://github.com/user-attachments/assets/df5e69ee-18f2-4ddc-b387-7fc8b8123e4c" />

I think this section could be called something like `A New Look at Language Choices`

Author: Thanks, I like this suggestion

NNLS~\cite{Lawson1995} for optimization;
ARPACK-NG~\cite{Lehoucq2023ARPACKNG} and PROPACK~\cite{Larsen1998} for sparse eigenvalue problems
and singular value decompositions respectively; AMOS~\cite{Amos1986} and
SPECFUN~\cite{Zhang1997} for special functions;
CDFLIB~\cite{Brown1994CDFLIB}, mvndist~\cite{Genz1992}, and
statlib~\cite{Dinneen1976,Best1975,Royston1995,Maclaren1985} for probability
distributions and statistical functions;
and FFTPACK~\cite{Swarztrauber1982} for fast Fourier transforms.

Despite the enormous utility these codes have provided, they also presented a number
of challenges on the road to SciPy 2.0. A number of the above codes are not reentrant
meaning that they cannot be safely called from multiple threads simultaneously.
Previously, SciPy has used locking to prevent this. However, the introduction of
free-threaded CPython 3.13, and its subsequent maturation in CPython 3.14, which for the
first time allows multiple threads
to execute Python code concurrently without the Global
Interpreter Lock (GIL)~\cite{gross2023pep703}. Unfortunately,
users of these non-reentrant codes would not be able to take advantage of this

> new feature. Likewise, another exciting development SciPy wished to support was

Author: @rgommers it would be great if you could clarify the benefits of the Fortran rewrites for free threading. I have done my best to convey my understanding but it is a little out of my wheel house

Reviewer 2: I think I could write something reasonable for this if @rgommers doesn't have bandwidth at the moment

Author: Either would be great :) Likewise, feel free to make changes to other parts - I am not precious about what I have written here, I just wanted to start getting words on paper.

\section{Introduction}

\section{A New Look at Language Choices}

As discussed in~\cite{Virtanen2020}, much of SciPy's original success can be attributed

> to the convenient Python interfaces it provides to popular compiled numerical libraries

Reviewer 3: ```suggestion
to the convenient Python interfaces it provided to popular compiled numerical libraries
```

that were otherwise distributed primarily as standalone Fortran source
packages, such as those available through the Netlib repository~\cite{netlib}.

> This includes:

Reviewer 3: ```suggestion
This included:
```

and singular value decompositions respectively; AMOS~\cite{Amos1986} and
SPECFUN~\cite{Zhang1997} for special functions;
CDFLIB~\cite{Brown1994CDFLIB}, mvndist~\cite{Genz1992}, and
statlib~\cite{Dinneen1976,Best1975,Royston1995,Maclaren1985} for probability
distributions and statistical functions;
and FFTPACK~\cite{Swarztrauber1982} for fast Fourier transforms.
Consequently, SciPy 1.0 consisted of approximately $50\%$ Python, $25\%$ Fortran,
$20\%$ C, $3\%$ Cython, and $2\%$ C++ code~\cite{Virtanen2020}.

Despite the enormous utility these codes have provided, they also presented a number
of challenges on the road to SciPy 2.0. By far the most significant of these was
limited availability of Fortran compilers for Windows and
emerging new platforms such as WebAssembly~\cite{Haas2017}. This left SciPy
in a precarious position where distributing SciPy was challenging and required
a custom MinGW toolchain to build SciPy on Windows~\cite{obermeier2023eu}.
While tools such as \texttt{flang-new} (now \texttt{flang})
and \texttt{LFortran} showed promise in addressing this issue, they were not
mature enough to be adopted by SciPy at the time and waiting for them to mature was not a

> viable approach for something so integral to the project's future.

Reviewer 3: I might instead say something like 'was not deemed a viable approach'

In addition, there were a number of further issues pertaining to the Fortran 77
implementations of these codes. A number of them were not reentrant,

> meaning that they could not be safely called from multiple threads simultaneously.

Reviewer 4: Here terminology got a bit mixed up I think. Re-entrant means typically you call a function A that interacts with a function B but B also calls A in its body. Now the function A has to stop wherever it is and rewind itself back up and start over with no side effects.

Thread-safe is a similar concept but that means it does not have any internal state at any time so when Thread A calls function X and Thread B calls function X there are no side effects to any of them irrespective of the timing of the calls.

There are some technical differences but I don't think any of us need to study them to make a point. What happened is that I removed all internal state and local storage variables.

Author: Oops, thanks

Previously, SciPy used locking to prevent this. However, the introduction of
free-threading in CPython 3.13, and its subsequent maturation in CPython 3.14, meant
that, for the first time, multiple threads could execute Python code concurrently
without the Global Interpreter Lock (GIL)~\cite{gross2023pep703}. Unfortunately,
users of these non-reentrant codes would not be able to take advantage of this
new feature.

> Other challenges presented by these codes include a lack of support for

Reviewer 3: ```suggestion
Other challenges presented by these codes included a lack of support for
```

64-bit integer types (ILP64) and lack of control over the random number generator used.
Finally, many of these codes are no longer actively maintained, meaning that the
SciPy developers were effectively responsible for fixing any bugs that arose.

> However, being a Python library, few members of the SciPy developer had expertise

Reviewer 5: ```suggestion
However, being a Python library, few members of the SciPy developer community had expertise
```

to tackle these issues. This often led to bugs remaining unresolved indefinitely.

After careful evaluation of the above issues, it was decided that for SciPy 2.0, these

> codes needed either to be replaced with a

Reviewer 5: ```suggestion
codes needed either to be replaced with 
```

> more recent, actively maintained alternatives or to be translated from their

Reviewer 3: ```suggestion
more recent, actively maintained alternatives, or to be translated from their
```

original Fortran 77 source code to either pure Python,
Cython~\cite{Behnel2011}, or C/C++ in order to secure the long term future of the
project. Consequently, COBYLA has been replaced with
a Python implementation from the PRIMA~\cite{Zhang_2023} package, FFTPACK has been replaced with

> ducc.\@fft~\cite{reinecke2019ducc}, and some of the special and statistical functions

Reviewer 3: a comment in passing: cc @mreineck, it would be nice if you would like to discuss the development of `pocketfft` and `ducc.fff` briefly in the paper!

Author: I think we could briefly comment on the improvements from FFTPACK but IMO the actual development of these libraries beyond the scope of this paper (not because I don't think `pocketfft`/`ducc.fff` are great pieces of work, they are, but I worry about the scope ballooning too much)

Reviewer 3: yeah, I suppose I meant more a brief mention of what motivated SciPy to switch or what the most tangible benefits for users are

Author: Although on second thought if we have dedicated section on the new fft module then I agree some more detail about things a user would care about could be nice. (Sorry I'd already forgotten what I wrote in my own plan!)

Reviewer 6: I'm not totally sure either whether this should go into the paper, but I'm happy to add 1-2 paragraphs if there is consensus that this would be a good thing.
(But the brutally honest answer is: I was absolutely fascinated by FFTPACK, but annoyed that it wasn't fully compliant Fortran 77, and I had a bit too much time on my hands ...)
Documenting some of the `pocketfft` and `ducc` design decisions is something I hope to do at some point, but probably not as a formal publication.

> mentioned have been replaced with counterparts from Boost.\@Math. Translation of the remaining

Reviewer 3: citation?

Author: good point! https://github.com/boostorg/math/issues/1433

codes posed a significant undertaking, requiring over \todo{\#} lines
of code to be translated by hand. The completion of this work has been a major milestone for

> SciPy 2.0 leading to the resolution of the challenges described above as well as many

Reviewer 3: ```suggestion
SciPy 2.0, leading to the resolution of the challenges described above as well as many
```

long-standing bugs. Furthermore, these translations provide the foundation for
future development and improvements, for example, improved special functions,

> see \todo{Forward Ref} for more details. These developments would not have originally

Reviewer 3: ```suggestion
future development and improvements, for example, improved special functions — 
see \todo{Forward Ref} for more details. These developments would not have originally
```

> been possible due to the lack of expertise in Fortran 77 within the SciPy developer community.

Reviewer 5: We're certainly not the only ones with such a concern regarding expertise. There's a DOE whitepaper (https://doi.org/10.2172/1970284) expressing similar concerns:

> This document examines risks associated with relying on Fortran for codes critical to nuclear security into
the late 2030s

> It is very likely that we will be unable to staff Fortran projects with top-rate computer scientists and
computer engineers.

> It is very likely that Fortran will preclude effective use of important advances in computing
technology.

Not sure I'd cite that on its own, although if we had a balanced view (i.e., maybe also a European institution report or something), might be interesting to note this is a broader concern not isolated to the Python community.

> been possible due to the lack of expertise in Fortran 77 within the SciPy developer community.

Reviewer 3: ```suggestion
see \todo{Forward Ref} for more details. Such developments were previously infeasible
due to the lack of expertise in Fortran 77 within the SciPy developer community.
```
